roles/matrix-nginx-proxy/templates/nginx/conf.d/matrix-synapse.conf.j2

```diff
@@ -159,6 +226,50 @@
 	}
 {% endmacro %}
 
+{% set generic_workers = matrix_nginx_proxy_synapse_workers_list|selectattr('type', 'equalto', 'generic_worker')|list %}
+{% set media_repository_workers = matrix_nginx_proxy_synapse_workers_list|selectattr('type', 'equalto', 'media_repository')|list %}
+{% set user_dir_workers = matrix_nginx_proxy_synapse_workers_list|selectattr('type', 'equalto', 'user_dir')|list %}
+{% set frontend_proxy_workers = matrix_nginx_proxy_synapse_workers_list|selectattr('type', 'equalto', 'frontend_proxy')|list %}
+{% if matrix_nginx_proxy_synapse_workers_enabled %}
+	# Round Robin "upstream" pools for workers
+
+	{% if generic_workers %}
+	upstream generic_worker_upstream {
+		# ensures that requests from the same client will always be passed
+		# to the same server (except when this server is unavailable)
+		ip_hash;
```

Comment:
No. Which `key` for the http://nginx.org/en/docs/http/ngx_http_upstream_module.html#hash directive would you propose for that?
 The average user might have two devices online with different IPs. Balancing by IP would get the traffic of both be handled by different workers, balancing by authorization would make them go to the same.
It might marginally improve the load because of less cache faults.. Than again, no idea how well the new `redis` IPC will catches that.

Reply:
`hash $http_authorization` should balance connections by the value of the Authorization header. For a single-user multi-device setups, there is little benefit, but if you have e.g. 10 users with each 2-3 sessions, each user w/ all their sessions would be bound to one backend, enabling fast `/sync`s and not stalling other users too much if someone does an `/initialsync` - this applies only to the `/[initial]sync` endpoints (and 2-3 others), the rest of the CS-API-endpoints can be balanced by source IP aswell.

Reply:
How would you make all of the user's sessions land on the same worker? Doesn't each session gets its own access token and thus has its own unique `Authorization` header.

The best that should be possible with this is to have each session land the same worker, but not have all these sessions (for the same user) do that.

Unless there's something smart which could parse the macaroon out of the `Authorization` header, read the user id out of it and then balance based on that value.

I see that `workers.md` says the same:

> For /sync and /initialSync requests it will be more efficient if all requests from a particular user are routed to a single instance. Extracting a user ID from the access token or Authorization header is currently left as an exercise for the reader.

Well, how does one solve that exercise (in nginx), I don't know.

Reply:
https://github.com/bungle/awesome-resty#authentication-and-authorization has a macroon lib, so in principle it might be possible using OpenResty (nginx distribution with lots of lua modules)